### ISSUE TITLE
Case class decoding and empty cells

### ISSUE THREAD

**Issue by csokol** [nrinaudo/kantan.csv#88]
I'm facing a weird issue when decoding CSV files using case classes support. When the cell is an empty string, the decoding fails with an non descriptive message.

The same CSV is successfully parsed with tuples or (which is even stranger) in case the input ends with a line break.

Here's how to reproduce:

```scala
import kantan.csv.CsvConfiguration._
import kantan.csv._
import kantan.csv.ops._
import kantan.csv.generic._

final case class Order(id: String, product: String)

val csvWithEmptyField = """"Order ID";"Product"
"1";"P1"
"1";"""""

val csvWithoutEmptyField = """"Order ID";"Product"
"1";"P1"
"1";"P2""""


def parseAndPrint(txt: String) = {
  val reader = txt.asCsvReader[Order](CsvConfiguration(';', '"', QuotePolicy.Always, Header.Implicit))
  reader.foreach(println)
}

def parseAndPrintWithTuple(txt: String) = {
  type Tup = (String, String)
  val reader = txt.asCsvReader[Tup](CsvConfiguration(';', '"', QuotePolicy.Always, Header.Implicit))
  reader.foreach(println)
}

parseAndPrint(csvWithoutEmptyField) //works
println("-----\n")
parseAndPrint(csvWithEmptyField)  //fails
println("-----\n")
parseAndPrint(csvWithEmptyField + "\n") //works
println("-----\n")
parseAndPrintWithTuple(csvWithEmptyField) //works
println("-----\n")
parseAndPrintWithTuple(csvWithEmptyField + "\n") //works
```

The output of this code is as follows:

```
Success(Order(1,P1))
Success(Order(1,P2))
-----

Success(Order(1,P1))
Failure(OutOfBounds: 0 is not a valid index)
-----

Success(Order(1,P1))
Success(Order(1,))
-----

Success((1,P1))
Success((1,))
-----

Success((1,P1))
Success((1,))

```

Am I doing something wrong or is this a bug?

**Comment by nrinaudo** [comment]
The problem here, which you had no way of guessing, is because you're using generic derivation, which is really strict about its inputs. Your content *without* a line break at the end, according to the RFC, terminates after `1` - its last row contains a single cell. Generic derivation needs exact matches, hence your error.

It's easy enough to work around though, write your own decoder: 
```scala
scala> implicit val orderDecoder: RowDecoder[Order] = RowDecoder.ordered(Order.apply _)

scala> csvWithEmptyField.asCsvReader[Order](rfc.withCellSeparator(';').withHeader).foreach(println _)
Success(Order(1,P1))
Success(Order(1,))
```

**Comment by csokol** [comment]
It worked perfectly, thanks! Maybe we should add a word about that in the generic docs? Or is it already there and I missed it?

**Comment by nrinaudo** [comment]
I think I'll try to see if I can sort that out at the parser level first, but yes, you're right. It's just taht, contrary to what one my feel, csv doesn't lend itself all that well to generic derivation...

**Comment by nrinaudo** [comment]
@csokol this... is really, really odd. You seem to have find a bug in a weird corner case. I'll need to fix that. In the meantime, my first explanation was completely incorrect. Your problem comes from the fact that the last cell is `""`, which seems to break the internal parser. CSV, man.

**Comment by nrinaudo** [comment]
@csokol yes, that was a bug. Thanks for spotting it. It's now fixed. Do you have a urgent need for this, or is the workaround of declaring a decoder manually enough to get you by?

**Comment by csokol** [comment]
Cool, I'm glad I've helped finding that weird corner case :)

The workaround is enough for now, don't need to worry with a release. Thanks!

**Comment by m3nadav** [comment]
Hey, I came across the same problem where last field of the row is missing if it's value was empty quotes (`,""`).
I try to decode a csv `ByteString` row into a `List[String]` using
`
      val rawRow = """field1,field2,"""""
      val retVal = rawRow.asCsvReader[List[String]](rfc).next match {
        case Result.Success(parsedLine)               => parsedLine
        case Result.Failure(ParseError.NoSuchElement) => List()
        case Result.Failure(e)                        => throw e
      }

      retVal.head shouldEqual "field1"
      retVal(1) shouldEqual "field2"
      retVal(2) shouldEqual ""
`
as you can see the last field should be empty string but that test fails because there is no 3rd field in the result.

Is this related? if so I could really use the new release :D

**Comment by nrinaudo** [comment]
Running that on my local version, I get:

```scala
scala>  val rawRow = """field1,field2,"""""
rawRow: String = field1,field2,""

scala>val retVal = rawRow.asCsvReader[List[String]](rfc).next match {
     |         case Success(parsedLine)               => parsedLine
     |         case Failure(ParseError.NoSuchElement) => List()
     |         case Failure(e)                        => throw e
     |       }
retVal: List[String] = List(field1, field2, "")
```

If you confirm that this is the expected output, I will try and cut a release soon.

**Comment by nrinaudo** [comment]
Fixed in 0.2.1
